src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,529 @@
+ *   are intended to track the relation accesses within a transaction. The
+ *   logic here is mostly useful when a reference table is referred by
+ *   a distributed table via a foreign key. Whenever such a pair of tables
+ *   are acccesed inside a transaction, Citus should detect and act
+ *   accordingly.
+ *
+ * Copyright (c) 2018, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "miscadmin.h"
+
+#include "access/xact.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/relation_access_tracking.h"
+#include "utils/hsearch.h"
+
+
+#define PARALLEL_MODE_FLAG_OFFSET 3
+
+/*
+ * Hash table mapping relations to the
+ *      (relationId) = (relationAccessType and relationAccessMode)
+ *
+ * RelationAccessHash is used to keep track of relation accesses types (e.g., select,
+ * dml or ddl) along with access modes (e.g., no access, sequential access or
+ * parallel access).
+ *
+ * We keep an integer per relation and use some of the bits to identify the access types
+ * and access modes.
+ *
+ * We store the access types in the first 3 bits:
+ *  - 0th bit is set for SELECT accesses to a relation
+ *  - 1st bit is set for DML accesses to a relation
+ *  - 2nd bit is set for DDL accesses to a relation
+ *
+ * and, access modes in the next 3 bits:
+ *  - 3rd bit is set for PARALLEL SELECT accesses to a relation
+ *  - 4th bit is set for PARALLEL DML accesses to a relation
+ *  - 5th bit is set for PARALLEL DDL accesses to a relation
+ *
+ */
+typedef struct RelationAccessHashKey
+{
+	Oid relationId;
+} RelationAccessHashKey;
+
+typedef struct RelationAccessHashEntry
+{
+	RelationAccessHashKey key;
+
+	int relationAccessMode;
+} RelationAccessHashEntry;
+
+static HTAB *RelationAccessHash;
+
+
+static void RecordPlacementAccessToCache(Oid relationId,
+										 ShardPlacementAccessType accessType);
+static RelationAccessMode GetRelationAccessMode(Oid relationId,
+												ShardPlacementAccessType accessType);
+static void RecordParallelRelationAccess(Oid relationId, ShardPlacementAccessType
+										 placementAccess);
+static void RecordParallelRelationAccessToCache(Oid relationId,
+												ShardPlacementAccessType placementAccess);
+
+
+/*
+ * Empty RelationAccessHash, without destroying the hash table itself.
+ */
+void
+ResetRelationAccessHash()
+{
+	hash_delete_all(RelationAccessHash);
+}
+
+
+/*
+ * Allocate RelationAccessHash.
+ */
+void
+AllocateRelationAccessHash()
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(RelationAccessHashKey);
+	info.entrysize = sizeof(RelationAccessHashEntry);
+	info.hash = tag_hash;
+	info.hcxt = ConnectionContext;
```

Comment:
Optional: It might be slightly safer to recreate the hash table for every transaction in a transaction context. It's not necessarily slower because resetting a memory context is faster than deleting individual elements.